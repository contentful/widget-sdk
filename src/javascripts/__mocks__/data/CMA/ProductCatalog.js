export const getSpaceFeature = jest.fn().mockResolvedValue(true);
export const FEATURES = {};
