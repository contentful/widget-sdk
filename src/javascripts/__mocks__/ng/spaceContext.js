export const user = {
  email: 'someone-from@contenful.com'
};

export const users = {
  get: jest.fn().mockResolvedValue({})
};

export const organization = {
  sys: {
    id: 'org'
  }
};

export const cma = {
  createExtension: jest.fn().mockResolvedValue({}),
  deleteExtension: jest.fn().mockResolvedValue({}),
  getExtensions: jest.fn().mockResolvedValue({ items: [] }),
  getExtension: jest.fn().mockResolvedValue({})
};

export const webhookRepo = {
  getAll: jest.fn().mockResolvedValue([]),
  get: jest.fn().mockResolvedValue({}),
  logs: {
    getCall: jest.fn().mockResolvedValue({}),
    getHealth: jest.fn().mockResolvedValue({
      calls: {
        health: 50,
        total: 100
      }
    })
  }
};

export const publishedCTs = {
  getAllBare: jest.fn().mockReturnValue([])
};

export const widgets = {
  refresh: jest.fn().mockResolvedValue([])
};

export const contentPreview = {
  replaceVariablesInUrl: jest.fn(),
  getForContentType: jest.fn(),
  getSelected: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  remove: jest.fn()
};

export const getData = jest.fn();

export const netlifyAppConfig = {
  get: jest.fn().mockResolvedValue({})
};

export const endpoint = jest.fn();

export const getEnvironmentId = jest.fn();
