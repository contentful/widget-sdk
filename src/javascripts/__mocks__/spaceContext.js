import sinon from 'sinon';

export default {
  cma: {
    createExtension: sinon.stub().resolves({}),
    deleteExtension: sinon.stub().resolves({}),
    getExtension: sinon.stub().resolves({})
  },
  webhookRepo: {
    getAll: sinon.stub().resolves([]),
    get: sinon.stub().resolves({}),
    logs: {
      getCall: sinon.stub().resolves({}),
      getHealth: sinon.stub().resolves({})
    }
  },
  publishedCTs: {
    getAllBare: sinon.stub().returns([])
  },
  widgets: {
    refresh: sinon.stub().resolves([])
  },
  getData: sinon.stub()
};
