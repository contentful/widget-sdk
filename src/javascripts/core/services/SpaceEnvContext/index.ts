export { useContentTypes } from './useContentTypes';
export { useSpaceEnvContext } from './useSpaceEnvContext';
