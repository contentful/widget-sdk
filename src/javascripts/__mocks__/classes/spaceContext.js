import * as SpaceContext from '../ng/spaceContext';

export const getSpaceContext = jest.fn().mockReturnValue(SpaceContext);
