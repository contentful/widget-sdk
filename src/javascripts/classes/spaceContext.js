import { registerFactory } from 'NgRegistry';
import _ from 'lodash';
import { create as createViewMigrator } from 'saved-views-migrator';
import * as K from 'utils/kefir';
import { deepFreeze, deepFreezeClone } from 'utils/Freeze';
import { purgeContentPreviewCache } from 'services/contentPreview';
import { purgeApiKeyRepoCache } from 'app/settings/api/services/ApiKeyRepoInstance';

const MASTER_ENVIRONMENT_ID = 'master';

export default function register() {
  /**
   * @ngdoc service
   * @name spaceContext
   *
   * @description
   * This service holds all context related to a space, including
   * contentTypes, users, and helper methods.
   */
  registerFactory('spaceContext', [
    '$rootScope',
    function spaceContext($rootScope) {
      const publishedCTsBus$ = K.createPropertyBus([]);

      // Enforcements deinitialization function, when changing space
      let enforcementsDeInit;

      let ShareJSConnection;
      let shouldUseEnvEndpoint;
      let APIClient;
      let Telemetry;
      let createUserCache;
      let TheLocaleStore;
      let createSpaceMembersRepo;
      let createEnvironmentsRepo;
      let createLocaleRepo;
      let createUiConfigStore;
      let createSpaceEndpoint;
      let createPubSubClientForSpace;
      let PublishedCTRepo;
      let MembershipRepo;
      let accessChecker;
      let DocumentPool;
      let EnforcementsService;
      let TokenStore;
      let Auth;
      let Config;
      let client;

      const spaceContext = {
        async init() {
          [
            ShareJSConnection,
            { default: shouldUseEnvEndpoint },
            { default: APIClient },
            Telemetry,
            { default: createUserCache },
            { default: TheLocaleStore },

            { default: createSpaceMembersRepo },
            { create: createEnvironmentsRepo },
            { default: createLocaleRepo },
            { default: createUiConfigStore },
            { createSpaceEndpoint },
            PublishedCTRepo,
            MembershipRepo,
            accessChecker,
            DocumentPool,
            EnforcementsService,
            TokenStore,
            Auth,
            Config,
            { default: client },
            { createPubSubClientForSpace },
          ] = await Promise.all([
            import(/* webpackMode: "eager" */ 'data/sharejs/Connection'),
            import(/* webpackMode: "eager" */ 'data/shouldUseEnvEndpoint'),
            import(/* webpackMode: "eager" */ 'data/APIClient'),
            import(/* webpackMode: "eager" */ 'i13n/Telemetry'),
            import(/* webpackMode: "eager" */ 'data/userCache'),
            import(/* webpackMode: "eager" */ 'services/localeStore'),
            import(/* webpackMode: "eager" */ 'data/CMA/SpaceMembersRepo'),
            import(/* webpackMode: "eager" */ 'data/CMA/SpaceEnvironmentsRepo'),
            import(/* webpackMode: "eager" */ 'data/CMA/LocaleRepo'),
            import(/* webpackMode: "eager" */ 'data/UiConfig/Store'),
            import(/* webpackMode: "eager" */ 'data/Endpoint'),
            import(/* webpackMode: "eager" */ 'data/ContentTypeRepo/Published'),
            import(/* webpackMode: "eager" */ 'access_control/SpaceMembershipRepository'),
            import(/* webpackMode: "eager" */ 'access_control/AccessChecker'),
            import(/* webpackMode: "eager" */ 'data/sharejs/DocumentPool'),
            import(/* webpackMode: "eager" */ 'services/EnforcementsService'),
            import(/* webpackMode: "eager" */ 'services/TokenStore'),
            import(/* webpackMode: "eager" */ 'Authentication'),
            import(/* webpackMode: "eager" */ 'Config'),
            import(/* webpackMode: "eager" */ 'services/client'),
            import(/* webpackMode: "eager" */ 'services/PubSubService'),
          ]);
        },

        /**
         * @type {Published}
         * @description
         * A property containing data on the published CTs in the current space and space-specific repository methods.
         * Initialized with a predefined items$ bus to keep the bus reference the same between space switches.
         * It is updated in resetWithSpace method once we know we are in the context of a space.
         */
        publishedCTs: {
          items$: publishedCTsBus$.property,
        },
        /**
         * @ngdoc method
         * @name spaceContext#purge
         * @description
         * This method purges a space context, so it doesn't contain space any longer
         */
        purge: function () {
          resetMembers(spaceContext);
        },
        /**
         * @ngdoc method
         * @name spaceContext#resetWithSpace
         * @description
         * This method resets a space context with a given space.
         * It requires an API space object. Internally we create
         * @contentful/client instance.
         *
         * The returned promise resolves when all additional space
         * resources have been fetched:
         * - Extensions
         * - Content types
         * - UI Configs (space and user)
         * - Locales
         *
         * @param {API.Space} spaceData
         * @param {string?} uriEnvOrAliasId environment id based on the uri
         * @returns {Promise<self>}
         */
        resetWithSpace: async function (spaceData, uriEnvOrAliasId) {
          const self = this;
          accessChecker.setSpace(spaceData);

          // `space` is @contentful/client.Space instance!
          let space = client.newSpace(spaceData);

          if (uriEnvOrAliasId) {
            // creates env aware routes on space
            space = space.makeEnvironment(uriEnvOrAliasId, shouldUseEnvEndpoint);
          }

          const spaceId = space.getId();

          self.endpoint = createSpaceEndpoint(Config.apiUrl(), spaceId, Auth, uriEnvOrAliasId);

          resetMembers(self);
          self.space = space;
          self.cma = new APIClient(self.endpoint);
          self.users = createUserCache(self.endpoint);
          self.localeRepo = createLocaleRepo(self.endpoint);
          self.organization = deepFreezeClone(self.getData('organization'));

          // TODO: publicly accessible docConnection is
          // used only in a process of creating space out
          // of a template. We shouldn't use it in newly
          // created code.

          self.docConnection = ShareJSConnection.create(
            Config.otUrl,
            Auth,
            spaceId,
            uriEnvOrAliasId || MASTER_ENVIRONMENT_ID
          );

          self.memberships = MembershipRepo.create(self.endpoint);
          self.members = createSpaceMembersRepo(self.endpoint);
          self.user = K.getValue(TokenStore.user$);

          // This happens here, rather than in `prelude.js`, since it's scoped to a space
          // and not the user, so the spaceId is required.
          enforcementsDeInit = EnforcementsService.init(spaceId);

          const start = Date.now();
          return Promise.all([
            setupEnvironments(self, uriEnvOrAliasId),
            TheLocaleStore.init(self.localeRepo),
            setupPublishedCTsBus(self).then(() => {
              const ctMap = self.publishedCTs
                .getAllBare()
                .reduce((acc, ct) => ({ ...acc, [ct.sys.id]: ct }), {});
              self.uiConfig = createUiConfigStore(
                space,
                self.endpoint,
                self.publishedCTs,
                createViewMigrator(ctMap)
              );
            }),
            (async () => {
              self.pubsubClient = await createPubSubClientForSpace(spaceId);
            })(),
            (async () => {
              self.docPool = await DocumentPool.create(
                self.docConnection,
                self.endpoint,
                self.organization.sys.id,
                spaceId
              );
            })(),
          ]).then(() => {
            Telemetry.record('space_context_http_time', Date.now() - start);
            // TODO: remove this after we have store with combined reducers on top level
            // string is hardcoded because this code _is_ temporary
            $rootScope.$broadcast('spaceContextUpdated');
            return self;
          });
        },

        /**
         * @description
         * Returns ID of current space, if set
         * @returns String
         */
        getId: function () {
          return this.space && this.space.getId();
        },

        /**
         * @description
         * Returns current space, if set
         * @returns Object
         */
        getSpace: function () {
          return this.space;
        },

        /**
         * @name spaceContext#getEnvironmentId
         * @description
         * Returns the current environment ID, defaulting to `master`.
         * Returns `undefined` if no space is set.
         * @returns string
         */
        getEnvironmentId: function () {
          if (this.space) {
            return _.get(this, ['space', 'environment', 'sys', 'id'], MASTER_ENVIRONMENT_ID);
          }
        },

        /**
         * @name spaceContext#isMasterEnvironment
         * @param {envOrAlias} Environment or Alias
         * @description
         * Returns whether the current environment is aliased to 'master'
         * or is 'master'
         * Returns true for an environment object with a re-written sys.id of
         * 'master' (aka the environment as accessed via it's alias)
         * @returns boolean
         */
        isMasterEnvironment: function (
          envOrAlias = _.get(this, ['space', 'environment'], { sys: { id: MASTER_ENVIRONMENT_ID } })
        ) {
          if (
            envOrAlias.sys.id === MASTER_ENVIRONMENT_ID ||
            _.find(envOrAlias.sys.aliases, (alias) => alias.sys.id === MASTER_ENVIRONMENT_ID)
          ) {
            return true;
          }
          return false;
        },

        /**
         * @name spaceContext#getAliasesIds
         * @description
         * Returns the ids of the environments aliases referencing the current environment
         * @returns array<string>
         */
        getAliasesIds: function (
          env = _.get(this, ['space', 'environment'], { sys: { id: MASTER_ENVIRONMENT_ID } })
        ) {
          if (!env.sys.aliases) return [];
          return env.sys.aliases.map(({ sys }) => sys.id);
        },

        /**

         /**
         * @name spaceContext#hasOptedIntoAliases
         * @description
         * Checks if the space is opted in to the environment alias feature
         * @returns boolean
         */
        hasOptedIntoAliases: function (environments = _.get(this, ['environments'], [])) {
          return environments.some(({ sys: { aliases = [] } }) => aliases.length > 0);
        },

        /**
         * @ngdoc method
         * @name spaceContext#getData
         * @param {string} path
         * @param {*} defaultValue
         * @description
         * Returns nested value stored under `path` in current `space.data`.
         * If not found, returns `defaultValue` (`undefined` when not provided)
         * @returns *
         */
        getData: function (path, defaultValue) {
          const data = _.get(this, 'space.data', {});
          return _.get(data, path, defaultValue);
        },
      };

      resetMembers(spaceContext);
      return spaceContext;

      /**
       * Extend the publishedCTs prop with the space-specific repo methods excluding
       * the repo items$ bus – to keep the same instance between space switches.
       * That means that calling on provided methods (like repo.publish()) will update the repo.items$, but not self.publishedCTs.items$.
       * So the additional listener on values changes is required to keep (self.publishedCTs.items$ = publishedCTsBus$) bus up-to-date.
       */
      async function setupPublishedCTsBus(spaceContext) {
        /** @type {Published} */
        const publishedCTsForSpace = PublishedCTRepo.create(spaceContext.space);
        _.assign(spaceContext.publishedCTs, _.omit(publishedCTsForSpace, 'items$'));
        await spaceContext.publishedCTs.refresh();
        // Synchronous first update since there's a current value now.
        K.onValue(publishedCTsForSpace.items$, (items) => publishedCTsBus$.set(items));
      }

      function resetMembers(spaceContext) {
        // Deinit the enforcement refreshing on space ID change, so that
        // the previous space ID enforcement information isn't queried
        if (enforcementsDeInit) {
          enforcementsDeInit();
        }

        spaceContext.uiConfig = null;
        spaceContext.space = null;
        spaceContext.users = null;

        purgeContentPreviewCache();
        purgeApiKeyRepoCache();

        if (spaceContext.docPool) {
          spaceContext.docPool.destroy();
          spaceContext.docPool = null;
        }
        if (spaceContext.docConnection) {
          spaceContext.docConnection.close();
          spaceContext.docConnection = null;
        }
      }

      /**
       * Setup the environments and environmentMeta and add it to the spaceContext
       *
       * @param {SpaceContext} spaceContext
       * @param {string} uriEnvOrAliasId
       * @returns {Promise}
       */
      function setupEnvironments(spaceContext, uriEnvOrAliasId = MASTER_ENVIRONMENT_ID) {
        return createEnvironmentsRepo(spaceContext.endpoint)
          .getAll()
          .then(({ environments, aliases = [] }) => {
            spaceContext.environments = deepFreeze(
              environments.sort(
                (envA, envB) =>
                  spaceContext.isMasterEnvironment(envB) - spaceContext.isMasterEnvironment(envA)
              )
            );
            spaceContext.aliases = deepFreeze(aliases);
          })
          .then(() => {
            let environment = spaceContext.environments.find(
              ({ sys }) => sys.id === uriEnvOrAliasId
            );

            if (!environment) {
              // the current environment is aliased
              environment = spaceContext.environments.find(({ sys: { aliases = [] } }) =>
                aliases.some(({ sys }) => sys.id === MASTER_ENVIRONMENT_ID)
              );
            }
            spaceContext.space.environment = environment;
          })
          .then(() => {
            const aliases = spaceContext.getAliasesIds(spaceContext.space.environment);
            spaceContext.space.environmentMeta = {
              environmentId: spaceContext.getEnvironmentId(),
              isMasterEnvironment: spaceContext.isMasterEnvironment(spaceContext.space.environment),
              aliasId: aliases[0], // for now we assume that there is only alias ('master')
              optedIn: spaceContext.hasOptedIntoAliases(),
            };
          });
      }
    },
  ]);
}
