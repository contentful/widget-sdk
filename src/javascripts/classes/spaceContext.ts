import { getModule } from 'core/NgRegistry';
import _ from 'lodash';
// @ts-expect-error missing type declarations
import { create as createViewMigrator } from 'saved-views-migrator';
import * as K from 'core/utils/kefir';
import { deepFreeze, deepFreezeClone } from 'utils/Freeze';
import { purgeContentPreviewCache } from 'features/content-preview';
import { purgeApiKeyRepoCache } from 'features/api-keys-management';

import shouldUseEnvEndpoint from 'data/shouldUseEnvEndpoint';
import APIClient from 'data/APIClient';
import * as Telemetry from 'i13n/Telemetry';
import createUserCache from 'data/userCache';
import TheLocaleStore from 'services/localeStore';
import createSpaceMembersRepo from 'data/CMA/SpaceMembersRepo';
import { create as createEnvironmentsRepo } from 'data/CMA/SpaceEnvironmentsRepo';
import { createLocaleRepo } from 'data/CMA/LocaleRepo';
import createUiConfigStore from 'data/UiConfig/Store';
import { createSpaceEndpoint } from 'data/Endpoint';
import * as PublishedCTRepo from 'data/ContentTypeRepo/Published';
import * as MembershipRepo from 'access_control/SpaceMembershipRepository';
import * as accessChecker from 'access_control/AccessChecker';
import * as DocumentPool from 'data/DocumentPool/DocumentPool';
import * as EnforcementsService from 'services/EnforcementsService';
import * as TokenStore from 'services/TokenStore';
import * as Auth from 'Authentication';
import * as Config from 'Config';
import client from 'services/client';
import { captureError } from 'core/monitoring';
import { createPubSubClientForSpace } from 'services/PubSubService';
import { getSpaceEnvCMAClient, getCMAClient } from 'core/services/usePlainCMAClient';
import { GlobalEventBus, GlobalEvents } from 'core/services/GlobalEventsBus';
import {
  EnvironmentProps as Environment,
  EnvironmentAliasProps as EnvironmentAlias,
  EnvironmentProps,
  ContentTypeProps,
} from 'contentful-management/types';
import { SpaceContextType } from './spaceContextTypes';
import createResourceService from 'services/ResourceService';

const MASTER_ENVIRONMENT_ID = 'master';

// Enforcements deinitialization function, when changing space
let enforcementsDeInit;

const spaceContext = initSpaceContext();
export const getSpaceContext = (): SpaceContextType => spaceContext;

// Util function to reset the spaceContext with spaceId (and environmentId) params
export const resetWithSpace = async (params: { spaceId: string; environmentId?: string }) => {
  const spaceData = await TokenStore.getSpace(params.spaceId);
  return spaceContext.resetWithSpace(spaceData, params.environmentId);
};

/**
 * This service holds all context related to a space, including
 * contentTypes, users, and helper methods.
 */
function initSpaceContext(): SpaceContextType {
  const publishedCTsBus$ = K.createPropertyBus<any>([]);

  GlobalEventBus.on(GlobalEvents.RefreshPublishedContentTypes, () => {
    if (spaceContext?.publishedCTs?.refresh) {
      spaceContext.publishedCTs.refresh();
    }
  });

  const context: SpaceContextType = {
    cma: null as unknown as SpaceContextType['cma'],
    endpoint: null as unknown as SpaceContextType['endpoint'],
    memberships: null as unknown as SpaceContextType['memberships'],
    members: null as unknown as SpaceContextType['members'],
    pubsubClient: null as unknown as SpaceContextType['pubsubClient'],
    user: null as unknown as SpaceContextType['user'],
    organization: null as unknown as SpaceContextType['organization'],
    aliases: [],
    environments: [],
    docPool: null,
    users: null,
    resources: null as unknown as SpaceContextType['resources'],
    uiConfig: null,
    space: null as unknown as SpaceContextType['space'],
    resettingSpace: false,

    /**
     * @type {Published}
     * @description
     * A property containing data on the published CTs in the current space and space-specific repository methods.
     * Initialized with a predefined items$ bus to keep the bus reference the same between space switches.
     * It is updated in resetWithSpace method once we know we are in the context of a space.
     */
    publishedCTs: {
      items$: publishedCTsBus$.property,
      refresh: () => Promise.resolve(undefined),
      getAllBare: () => [],
      get: () => null,
      publish: () => Promise.resolve({} as ContentTypeProps),
      unpublish: () => Promise.resolve({} as ContentTypeProps),
    },

    purge: function () {
      resetMembers(spaceContext);
    },

    resetWithSpace: async function (spaceData, uriEnvOrAliasId) {
      spaceContext.resettingSpace = true;

      try {
        accessChecker.setSpace(spaceData);

        /**
         * @deprecated
         * This is a legacy client instance which returns entities as `.data` and uses
         * `X-Contentful-Skip-Transformation` for CMA requests which exposes internal IDs.
         * Use e.g. `spaceContext.cma` or `sdk.space`  instead wherever possible.
         */
        // @ts-expect-error missing newSpace type declarations
        let space = client.newSpace(spaceData);

        if (uriEnvOrAliasId) {
          // creates env aware routes on space
          space = space.makeEnvironment(uriEnvOrAliasId, shouldUseEnvEndpoint);
        }

        const spaceId = spaceData.sys.id;

        spaceContext.endpoint = createSpaceEndpoint(
          Config.apiUrl(),
          spaceId,
          Auth,
          uriEnvOrAliasId
        );

        resetMembers(spaceContext);

        spaceContext.space = space;
        spaceContext.cma = new APIClient(spaceContext.endpoint);
        spaceContext.users = createUserCache(spaceContext.endpoint);
        spaceContext.organization = deepFreezeClone(spaceContext.getData('organization'));

        const cmaPlainClient = getCMAClient(
          {
            spaceId,
            environmentId: uriEnvOrAliasId || MASTER_ENVIRONMENT_ID,
          },
          {
            //batch client doesn't support headers param in
            //entry.get and asset.get, used by EntityRepo
            noBatch: true,
          }
        );

        const localeRepo = createLocaleRepo(cmaPlainClient);

        spaceContext.memberships = MembershipRepo.create(spaceContext.endpoint);
        spaceContext.members = createSpaceMembersRepo(spaceContext.endpoint);
        spaceContext.user = K.getValue(TokenStore.user$);

        // This happens here, rather than in `prelude.js`, since it's scoped to a space
        // and not the user, so the spaceId is required.
        enforcementsDeInit = EnforcementsService.init(spaceId);

        // @ts-expect-error supress error
        spaceContext.pubsubClient = await createPubSubClientForSpace(spaceId);

        const start = Date.now();

        await setupEnvironments(spaceContext, uriEnvOrAliasId);
        spaceContext.docPool = await DocumentPool.create(
          spaceContext.pubsubClient,
          spaceContext.organization.sys.id,
          spaceId,
          spaceContext.space.environment,
          cmaPlainClient,
          spaceContext.endpoint
        );
        // This should be initialized after `setupEnvironments`, so that
        // the environment is properly set on spaceContext.space
        spaceContext.resources = createResourceService(spaceContext.endpoint);

        await TheLocaleStore.init(localeRepo);

        // This has to be called after `setupEnvironments`, so that
        // the environment is properly set on spaceContext.space
        await setupPublishedCTsBus(spaceContext);

        const ctMap = spaceContext.publishedCTs
          .getAllBare()
          .reduce((acc, ct) => ({ ...acc, [ct.sys.id]: ct }), {});

        spaceContext.uiConfig = createUiConfigStore(
          space,
          spaceContext.endpoint,
          spaceContext.publishedCTs,
          createViewMigrator(ctMap)
        );
        Telemetry.record('space_context_http_time', Date.now() - start);
        // TODO: remove this after we have store with combined reducers on top level
        // string is hardcoded because this code _is_ temporary
        getModule('$rootScope').$broadcast('spaceContextUpdated');
      } catch (err) {
        captureError(err);
      } finally {
        spaceContext.resettingSpace = false;
      }
      return spaceContext;
    },

    getId: function () {
      return this.space && this.space.getId();
    },

    /**
     * @description
     * Returns current space, if set
     * @returns Object
     * @deprecated
     * This is a legacy @contentful/client instance which returns entities as `.data` and uses
     * `X-Contentful-Skip-Transformation` for CMA requests which exposes internal IDs.
     * Use e.g. `spaceContext.cma` or `sdk.space`  instead wherever possible.
     */
    getSpace: function () {
      return this.space;
    },

    getEnvironmentId: function () {
      if (this.space) {
        return _.get(this, ['space', 'environment', 'sys', 'id'], MASTER_ENVIRONMENT_ID);
      }
    },

    getAliasId: function () {
      if (this.space) {
        return _.get(this, ['space', 'environmentMeta', 'aliasId'], undefined);
      }
    },

    isMasterEnvironment: function (environment) {
      // @ts-expect-error ignore
      const envOrAlias: Pick<Environment, 'sys'> =
        environment ||
        _.get(this, ['space', 'environment'], { sys: { id: MASTER_ENVIRONMENT_ID } });

      if (
        envOrAlias.sys.id === MASTER_ENVIRONMENT_ID ||
        _.find(envOrAlias.sys.aliases, (alias) => alias.sys.id === MASTER_ENVIRONMENT_ID)
      ) {
        return true;
      }
      return false;
    },

    isMasterEnvironmentById: function (envId) {
      const envOrAlias = _.get(
        this,
        ['environments'],
        [{ sys: { id: MASTER_ENVIRONMENT_ID } }]
      ).find(({ sys }) => sys.id === envId);
      return this.isMasterEnvironment(envOrAlias as any);
    },

    getAliasesIds: function (environment) {
      // @ts-expect-error ignore
      const env: Pick<Environment, 'sys'> =
        environment ||
        _.get(this, ['space', 'environment'], { sys: { id: MASTER_ENVIRONMENT_ID } });
      if (!env.sys.aliases) return [];
      return env.sys.aliases.map(({ sys }) => sys.id);
    },

    hasOptedIntoAliases: function (envs) {
      const environments = envs || _.get(this, ['environments'], []);
      return environments.some(({ sys: { aliases = [] } }) => aliases.length > 0);
    },

    getData: function (path, defaultValue) {
      const data = _.get(this, 'space.data', {});
      return _.get(data, path, defaultValue);
    },
  };

  resetMembers(context);
  return context;

  /**
   * Extend the publishedCTs prop with the space-specific repo methods excluding
   * the repo items$ bus – to keep the same instance between space switches.
   * That means that calling on provided methods (like repo.publish()) will update the repo.items$, but not self.publishedCTs.items$.
   * So the additional listener on values changes is required to keep (self.publishedCTs.items$ = publishedCTsBus$) bus up-to-date.
   */
  async function setupPublishedCTsBus(spaceContext) {
    const cma = getSpaceEnvCMAClient();

    /** @type {Published} */
    const publishedCTsForSpace = PublishedCTRepo.create(cma);

    _.assign(spaceContext.publishedCTs, _.omit(publishedCTsForSpace, 'items$'));
    await spaceContext.publishedCTs.refresh();
    // Synchronous first update since there's a current value now.
    K.onValue(publishedCTsForSpace.items$, (items) => publishedCTsBus$.set(items));
  }

  function resetMembers(spaceContext: SpaceContextType) {
    // Deinit the enforcement refreshing on space ID change, so that
    // the previous space ID enforcement information isn't queried
    if (enforcementsDeInit) {
      enforcementsDeInit();
    }

    spaceContext.uiConfig = null;
    // @ts-expect-error can be nullable
    spaceContext.space = null;
    spaceContext.users = null;
    spaceContext.resources = null;

    purgeContentPreviewCache();
    purgeApiKeyRepoCache();

    if (spaceContext.docPool) {
      spaceContext.docPool.destroy();
      spaceContext.docPool = null;
    }
  }

  /**
   * Setup the environments and environmentMeta and add it to the spaceContext
   *
   * @param {SpaceContext} spaceContext
   * @param {string} uriEnvOrAliasId
   * @returns {Promise}
   */
  function setupEnvironments(
    spaceContext: SpaceContextType,
    uriEnvOrAliasId: string = MASTER_ENVIRONMENT_ID
  ) {
    return createEnvironmentsRepo(spaceContext.endpoint)
      .getAll()
      .then(
        ({
          environments,
          aliases = [],
        }: {
          environments: Environment[];
          aliases: EnvironmentAlias[];
        }) => {
          spaceContext.environments = deepFreeze(
            environments.sort((envA, envB) => {
              return spaceContext.isMasterEnvironment(envB) > spaceContext.isMasterEnvironment(envA)
                ? 1
                : -1;
            })
          );
          spaceContext.aliases = deepFreeze(aliases);
        }
      )
      .then(() => {
        spaceContext.space.environment = spaceContext.environments.find(
          ({ sys }) => sys.id === uriEnvOrAliasId
        ) as EnvironmentProps;

        if (!spaceContext.space.environment) {
          // the current environment is aliased
          spaceContext.space.environment = spaceContext.environments.find(
            ({ sys: { aliases = [] } }) => aliases.some(({ sys }) => sys.id === uriEnvOrAliasId)
          ) as EnvironmentProps;
          spaceContext.space.environmentMeta = {
            environmentId: spaceContext.getEnvironmentId() as string,
            isMasterEnvironment: spaceContext.isMasterEnvironment(spaceContext.space.environment),
            optedIn: spaceContext.hasOptedIntoAliases(),
            aliasId: uriEnvOrAliasId,
          };
        } else {
          spaceContext.space.environmentMeta = {
            environmentId: spaceContext.getEnvironmentId() as string,
            isMasterEnvironment: spaceContext.isMasterEnvironment(spaceContext.space.environment),
            optedIn: spaceContext.hasOptedIntoAliases(),
          };
        }
      });
  }
}
