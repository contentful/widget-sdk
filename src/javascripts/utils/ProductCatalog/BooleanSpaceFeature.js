import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { getSpaceFeature } from 'data/CMA/ProductCatalog';
import { SpaceEnvContext } from 'core/services/SpaceEnvContext/SpaceEnvContext';

/**
 * React wrapper around Product Catalog space features with boolean variation.
 *
 * @export
 * @class BooleanSpaceFeature
 * @extends {React.Component}
 */
export default class BooleanSpaceFeature extends React.Component {
  static propTypes = {
    spaceFeatureKey: PropTypes.string.isRequired,
    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]).isRequired,
  };

  static contextType = SpaceEnvContext;

  state = {
    currentVariation: undefined,
  };
  async componentDidMount() {
    const { currentSpaceId } = this.context;
    const currentVariation = await getSpaceFeature(
      currentSpaceId,
      this.props.spaceFeatureKey,
      false
    );

    if (this.isUnmounted) {
      return;
    }
    this.setState({
      currentVariation,
    });
  }
  componentWillUnmount() {
    this.isUnmounted = true;
  }
  render() {
    if (_.isFunction(this.props.children)) {
      return this.props.children({ currentVariation: this.state.currentVariation });
    } else {
      return this.state.currentVariation ? this.props.children : null;
    }
  }
}
