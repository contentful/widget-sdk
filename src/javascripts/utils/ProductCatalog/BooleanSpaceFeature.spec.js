import React from 'react';
import Enzyme from 'enzyme';
import 'jest-enzyme';
import BooleanSpaceFeature from './BooleanSpaceFeature';
import flushPromises from 'test/helpers/flushPromises';
import * as spaceContextMocked from 'ng/spaceContext';
import * as ProductCatalog from 'data/CMA/ProductCatalog';
import { SpaceEnvContextProvider } from 'core/services/SpaceEnvContext/SpaceEnvContext';

const spaceId = 'space-id';

jest.mock('data/CMA/ProductCatalog', () => ({ getSpaceFeature: jest.fn() }));

describe('BooleanSpaceFeature Component', () => {
  beforeEach(() => {
    ProductCatalog.getSpaceFeature.mockClear();
    spaceContextMocked.getId.mockReturnValue(spaceId);
  });
  it('renders children if feature is enabled', async () => {
    ProductCatalog.getSpaceFeature.mockResolvedValue(true);
    const wrapper = Enzyme.mount(
      <SpaceEnvContextProvider>
        <BooleanSpaceFeature spaceFeatureKey="ff">1</BooleanSpaceFeature>
      </SpaceEnvContextProvider>
    );

    await flushPromises();
    wrapper.update();

    expect(ProductCatalog.getSpaceFeature).toHaveBeenCalledWith(`fg5eidi9k2qp`, 'ff', false);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders null if feature is disabled', async () => {
    ProductCatalog.getSpaceFeature.mockResolvedValue(false);
    const wrapper = Enzyme.mount(
      <SpaceEnvContextProvider>
        <BooleanSpaceFeature spaceFeatureKey="ff">1</BooleanSpaceFeature>
      </SpaceEnvContextProvider>
    );

    await flushPromises();
    wrapper.update();

    expect(ProductCatalog.getSpaceFeature).toHaveBeenCalledWith(`fg5eidi9k2qp`, 'ff', false);
    expect(wrapper).toMatchSnapshot();
  });

  describe('render prop', () => {
    it('invokes child function with current variation', async () => {
      ProductCatalog.getSpaceFeature.mockResolvedValue(true);

      const renderProp = jest.fn().mockImplementation(() => 1);
      const wrapper = Enzyme.mount(
        <SpaceEnvContextProvider>
          <BooleanSpaceFeature spaceFeatureKey="ff">{renderProp}</BooleanSpaceFeature>
        </SpaceEnvContextProvider>
      );

      await flushPromises();
      wrapper.update();

      expect(renderProp).toHaveBeenCalledWith({ currentVariation: true });
    });
  });
});
