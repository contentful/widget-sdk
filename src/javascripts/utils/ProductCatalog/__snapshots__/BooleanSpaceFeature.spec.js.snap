// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanSpaceFeature Component renders children if feature is enabled 1`] = `
<SpaceEnvContextProvider>
  <BooleanSpaceFeature
    spaceFeatureKey="ff"
  >
    1
  </BooleanSpaceFeature>
</SpaceEnvContextProvider>
`;

exports[`BooleanSpaceFeature Component renders null if feature is disabled 1`] = `
<SpaceEnvContextProvider>
  <BooleanSpaceFeature
    spaceFeatureKey="ff"
  />
</SpaceEnvContextProvider>
`;
